# XGBoost

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
dataset = pd.read_csv('Data.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

In [3]:
dataset.head()

Sample code number  Clump Thickness  Uniformity of Cell Size  \
0             1000025                5                        1   
1             1002945                5                        4   
2             1015425                3                        1   
3             1016277                6                        8   
4             1017023                4                        1   

   Uniformity of Cell Shape  Marginal Adhesion  Single Epithelial Cell Size  \
0                         1                  1                            2   
1                         4                  5                            7   
2                         1                  1                            2   
3                         8                  1                            3   
4                         1                  3                            2   

   Bare Nuclei  Bland Chromatin  Normal Nucleoli  Mitoses  Class  
0            1                3                1        1      2  
1           10                3                2        1      2  
2            2                3                1        1      2  
3            4                3                7        1      2  
4            1                3                1        1      2

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### XGBoost 구현

In [5]:
#pip install xgboost
# xgboost 버전을 0.90으로 다운그레이드 시켜야한다. 반응변수에서 0,1을 원한다.

from xgboost import XGBClassifier # xgboost를 이용한 회귀 모델을 만드는 클래스 XGBRegressor

classifier = XGBClassifier()
classifier.fit(X_train, y_train)

XGBClassifier()

In [6]:
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[84  3]
 [ 0 50]]


0.9781021897810219

의사결정나무가 95.9%로 가장 높은 정확도를 보여줬으나, 


XGBoost의 정확도가 97.8%로 예측이 3개만 틀릴만큼 굉장히 높다.


테스트 세트가 편향될 의향이 있을 수 있으므로, k-겹 교차 검증을 통해 다른 테스트 세트의 정확도를 계산해본다. 

### k-겹 교차 검증

In [7]:
# 테스트 세트에서 운 좋게 정확도가 높은 결과값을 얻지 않았다는 것을 증명해보자.

from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(accuracies.mean()*100))
print("SD: {:.2f} %".format(accuracies.std()*100))

Accuracy: 96.53 %
SD: 2.07 %


96.53%로 역시 높은 정확도와 2%의 적은 편차를 보여준다.